torch/_functorch/_aot_autograd/functional_utils.py

```diff
@@ -191,7 +191,79 @@ def has_metadata_mutation(f_arg, arg, *, check_only_storage_mutation: bool):
         return has_metadata_mutation_
 
 
-def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires_grad):
+def gen_alias_from_base(
+    aliased_base_tensor,
+    target_meta_tensor,
+    target_requires_grad,
+    target_functional_tensor=None,
+):
+    # Patch the correct requires_grad field of the output tensor, depending on whether:
+    # (i) the reconstructed output (out) was came from a tensor that requires grad or not;
+    # and (ii) the concrete returned output does require grad or not.
+    def patch_requires_grad(out):
+        if aliased_base_tensor.requires_grad and not target_requires_grad:
+            out = out.detach()
+        elif not aliased_base_tensor.requires_grad and target_requires_grad:
+            out.requires_grad_(True)
+        return out
+
+    # If provided, use the target functional tensor for replaying the views.
+    #
+    # In summary, we use the fact that FunctionalTensorWrapper saves the view
+    # functions applied to itself (collected during functionalization) so as
+    # to replay them (view functions) on the aliased_base_tensor.
+    if target_functional_tensor is not None:
```

Comment:
When is this ever `None`?

I think the invariant here is that if the output we are regenerating has `OutputType.alias_of_input`, then `target_functional_tensor` must be set? If that's true, then it might be a bit clearer if we:

(1) Update this function to take in an `OutputAliasInfo`, and grab `OutputAliasInfo.functional_tensor` here

(2) add an assert in `OutputAliasInfo.__post_init__` that `functional_tensor` is always set when `OutputAliasInfo.output_type is OutputType.alias_of_input`

what do you think?

Reply:
Specifically, [here](https://github.com/pytorch/pytorch/blob/ae6f8d923ceb032d3e933507b071284ba7ec218d/torch/_functorch/_aot_autograd/runtime_wrappers.py#L694-L696). Can we get the corresponding output functional tensor for a synthetic base?

Reply:
Since I'm explicitly not supporting in-place views, I don't think this invariant is true.